File: src/edu/neu/musicId/wav/WaveDataReader.java

package edu.neu.musicId.wav;

import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collection;

import edu.neu.musicId.exception.InvalidFormatException;
import edu.neu.musicId.util.Utilities;

/**
 * This class wraps an {@link InputStream} to read a {@link WaveData} object
 * from the stream.
 * 
 * @author ckolek
 * @since 1.0
 */
public class WaveDataReader implements Closeable {
    private final InputStream inputStream;

    public WaveDataReader(InputStream inputStream) {
        this.inputStream = inputStream;
    }

    /**
     * Constructs a {@link WaveData} object from the data read from the
     * {@link InputStream} contained by this {@code WaveDataReader}.
     * 
     * @return the extracted {@code WaveData} object
     * @throws InvalidFormatException if the data in the stream is not in the
     *             WAVE format
     * @throws IOException if data cannot be read for any reason other than the
     *             end of the input stream, if the input stream has been closed,
     *             or if some other I/O error occurs
     * @since 1.0
     */
    public WaveData read() throws InvalidFormatException, IOException {
        byte[] chunkIdBuf = new byte[4];

        long length = inputStream.read(chunkIdBuf);

        if (length < chunkIdBuf.length) {
            throw new IOException("could not read chunk ID");
        }

        final String chunkId = new String(chunkIdBuf);

        final boolean isLittleEndian;
        if (chunkId.equals(WaveData.CHUNK_ID__RIFF)) {
            isLittleEndian = true;
        } else if (chunkId.equals(WaveData.CHUNK_ID__RIFX)) {
            isLittleEndian = false;
        } else {
            throw new IOException("invalid chunk ID: " + chunkId);
        }

        byte[] chunkSizeBuf = new byte[4];

        length = inputStream.read(chunkSizeBuf);

        if (length < chunkSizeBuf.length) {
            throw new IOException("could not read chunk size");
        }

        final long chunkSize = Utilities.toInt32(chunkSizeBuf, isLittleEndian);

        byte[] formatBuf = new byte[4];

        length = inputStream.read(formatBuf);

        if (length < formatBuf.length) {
            throw new IOException("could not read chunk format");
        }

        final String format = new String(formatBuf);

        if (!format.equals("WAVE")) {
            throw new InvalidFormatException("invalid format: " + format);
        }

        Collection<WaveData.Chunk> chunks = new ArrayList<WaveData.Chunk>();

        while (length < chunkSize) {
            WaveData.Chunk chunk = readChunk(isLittleEndian);

            chunks.add(chunk);

            length += 8 + chunk.length();
        }

        if (length < chunkSize) {
            throw new IOException("did not read " + chunkSize + " bytes of chunk");
        }

        return new WaveData(chunkId, isLittleEndian, chunkSize, format, chunks);
    }

    private WaveData.Chunk readChunk(boolean isLittleEndian) throws IOException {
        byte[] chunkIdBuf = new byte[4];

        long length = inputStream.read(chunkIdBuf);

        if (length < chunkIdBuf.length) {
            throw new IOException("could not read sub-chunk ID");
        }

        final String chunkId = new String(chunkIdBuf);

        byte[] chunkSizeBuf = new byte[4];

        length = inputStream.read(chunkSizeBuf);

        if (length < chunkSizeBuf.length) {
            throw new IOException("could not read sub-chunk size");
        }

        final long chunkSize = Utilities.toInt32(chunkSizeBuf, isLittleEndian);

        byte[] data = new byte[(int) chunkSize];

        length = inputStream.read(data);

        if (length < chunkSize) {
            throw new IOException("did not read " + chunkSize + " bytes of sub-chunk");
        }

        return new WaveData.Chunk(chunkId, data);
    }

    @Override
    public void close() throws IOException {
        inputStream.close();
    }
}

File: wave/waveDataReader.py
__author__ = 'ckolek'

import io
from waveData import WaveData
from util import utilities
from util.exceptions import InvalidFormatError

class WaveDataReader:
    """
    This class provides the mechanism to read a WaveData object from a stream of
     raw bytes.
    """

    def __init__(self, stream):
        """
        :param stream: an input stream
        """

        self.stream = stream

    def read(self):
        """
        :return: A WaveDataObject read from the stream
        :raises: InvalidFormatError - If the data is not in a valid WAVE format
        """

        chunk_id_buf = self.stream.read(4)

        if chunk_id_buf is None or len(chunk_id_buf) < 4:
            raise InvalidFormatError('could not read chunk ID')

        try:
            chunk_id = chunk_id_buf.encode("ASCII")
        except Exception as e:
            raise InvalidFormatError(e.message)

        if chunk_id == 'RIFF':
            is_little_endian = True
        elif chunk_id == 'RIFX':
            is_little_endian = False
        else:
            raise InvalidFormatError('invalid chunk ID: ' + chunk_id)

        chunk_size_buf = self.stream.read(4)

        if chunk_size_buf is None or len(chunk_size_buf) < 4:
            raise InvalidFormatError('could not read chunk size')

        chunk_size = utilities.to_long(chunk_size_buf,
                                       is_unsigned=True,
                                       is_little_endian=is_little_endian)

        format_buf = self.stream.read(4)

        if format_buf is None or len(format_buf) < 4:
            raise InvalidFormatError('could not read format_buf')

        format = format_buf.encode("ASCII")

        if not format == 'WAVE':
            raise InvalidFormatError('invalid format: ' + format)

        length = 4

        chunks = set()

        while length < chunk_size:
            chunk = self.read_chunk(is_little_endian)

            length += 8 + chunk.length()

            chunks.add(chunk)

        return WaveData(chunk_id, is_little_endian, chunk_size, format, chunks)

    def read_chunk(self, is_little_endian):
        """
        Reads a WAVE subchunk from the input stream.

        :param is_little_endian: True if data is stored in little-endian format,
         False otherwise
        :return: the new Chunk object
        :raises: InvalidFormatError - If the data is not in a valid WAVE format
        """

        chunk_id_buf = self.stream.read(4)

        if chunk_id_buf is None or len(chunk_id_buf) < 4:
            raise InvalidFormatError('could not read sub-chunk ID')

        try:
            chunk_id = chunk_id_buf.encode("ASCII")
        except:
            raise InvalidFormatError('non-ASCII chunk ID')
        
        chunk_size_buf = self.stream.read(4)

        if chunk_size_buf is None or len(chunk_size_buf) < 4:
            raise InvalidFormatError('could not read sub-chunk size')

        chunk_size = utilities.to_long(chunk_size_buf,
                                       is_unsigned=True,
                                       is_little_endian=is_little_endian)

        data = self.stream.read(chunk_size)

        if data is None or len(data) < chunk_size:
            raise InvalidFormatError('did not read ' + `chunk_size` +
                                     ' bytes of sub-chunk')

        return WaveData.Chunk(chunk_id, data)

    def close(self):
        """
        Closes the underlying stream.
        """
        self.stream.close()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    def open(file_name):
        """
        Opens an input stream to read binary data from the file with the given
         name, and returns a WaveDataReader object which wraps the opened
         stream

        :param file_name: the name of the file to read from
        :return: the WaveDataReader object to read data from the file
        """
        return WaveDataReader(io.open(file_name, 'rb'))
    open = staticmethod(open)
